Smoke test two

taartspi opened the pull request:
Uses junit
SPI test works using linuxfs,  fffm needs to implement writeThenRead, I spoke Friday with Ncik. He was going to review recent PRs and help out with ffm.

Digittal output works with with gpiod, fails with ffm.  I will debug more to gather info and then open an issue.

Serial test not done.  I pasted some test code to change and use for this test, but nothing to review now.

taartspi committed:
First pass of the smoketest.  Missing the serial test, know what I want to try, no time tonight.  Need to test the ffm providers can be loaded, at the moment was using linuxfs and gpiod.  Documentation....  Decide on what logging method to use

taartspi committed:
First pass of the smoketest.  Missing the serial test, know what I want to try, no time tonight.  Need to test the ffm providers can be loaded, at the moment was using linuxfs and gpiod.  Documentation....  Decide on what logging method to use

taartspi committed:
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

stefanhaustein commented on pi4j-test/src/main/java/com/pi4j/test/README.md':
```diff
@@ -0,0 +1,64 @@
+
```
The file name seems to have a single quote at the end

stefanhaustein commented on pi4j-test/src/test/java/com/pi4j/test/smokeTest/SmokeTestControl.java:
```diff
@@ -0,0 +1,114 @@
+
+
+
+public class SmokeTestControl {
```
As this class has the actual test methods, I'd name this one "SmokeTest".

I am not sure about the other class. I find myself constantly scrolling back and forth to understand what's going on. 

Typically, tests are pretty self-contained.

So I'd lean towards just inlining it here for now.

If this becomes too much code in one file, I'd split into separate classes per type, "PwmSmokeTest", "SerialSmokeTest" etc.

I'd also lean towards limiting the scope of variables as much as possible. Can it be a local variable? Make it a local variable. Keeping stuff local provides assurance that it can't have an influence on other things going on elsewhere.

That said, having any smoke tests is a big step forward, so it's probably better to go ahead and check stuff. We can improve things any time.  Would be nice to name the class containing the @Test methods "...Test" though.

stefanhaustein commented on pi4j-test/src/test/java/com/pi4j/test/smokeTest/SmokeTest.java:
```diff
@@ -0,0 +1,315 @@
+
+
+public class SmokeTest {
+    static final String I2C_PROVIDER = "ffm-i2c"; //"linuxfs-i2c";
```
One way to organize this without manual changes would be as follows:

- Have an abstract base class "AbstractGpioSmokeTest" with a ctor that takes the provider and has all the gpio test methods
- For each provider, have a subclass "FfmGpioSmokeTest", "GpiodGpioSmokeTest" that just calls the super constructor with the provider but has no other methods

Another way might be parameterized tests.

taartspi committed:
Merge branch 'develop' into smokeTestTwo

taartspi committed:
Merge branch 'develop' into smokeTestTwo

taartspi committed:
Merge branch 'develop' into smokeTestTwo

taartspi committed:
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

taartspi committed:
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

stefanhaustein commented:
Looks good but please remove the trailing single quote in the file name of README.md

stefanhaustein commented on pi4j-test/src/test/java/com/pi4j/test/smokeTest/provider/ProviderSmokeTest.java:
```diff
@@ -0,0 +1,264 @@
+
+    @Test
+    public void testGpioIn() {
+          final int PIN_GPIO16 = 16;
```
indent

stefanhaustein commented on pi4j-test/src/test/java/com/pi4j/test/smokeTest/provider/ProviderSmokeTest.java:
```diff
@@ -0,0 +1,264 @@
+
+    @Test
+    public void testGpioOut() {
+        final int PIN_GPIO24 = 24;
```
If there is no "semantic" name for the constant and it's only used once, perhaps it would be more straightforward to inline it?

stefanhaustein commented on pi4j-test/src/test/java/com/pi4j/test/smokeTest/provider/ProviderSmokeTest.java:
```diff
@@ -0,0 +1,264 @@
+
+
+    private I2C  createI2cBMPDevice() {
+        final String I2C_PROVIDER = "ffm-i2c";
```
Sorry for causing confusion. Configuration constants should probably remain static final and at the top of the file. 

My concern about keeping stuff local was mostly about input/output instances etc.

stefanhaustein reviewed (approved):
Fine to go ahead as-is, but please fix the README file name.

I have added some optional nits to prove that I actually looked at the code :)

taartspi committed:
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

taartspi committed:
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

stefanhaustein commented:
Do you have any guess what's going wrong with the CI?

eitch commented:
I think this was just a fluke. Retrying

eitch reviewed (changes requested):
There seem to be some broken tests, and missing dependencies:

https://github.com/Pi4J/pi4j/actions/runs/19525975870/job/55933065003#step:4:7104

taartspi commented:
The test fails.  The serial test has not been completed. My work.  The spi fail has an issue against ffm to implement the writeThenRead    The gpioOut only fails on ffm and i am getting details to explain this in an issue.  I can put a sleep in the gpioOut test and it seems the thread sleeping lets some ffm code run properly     I could close the PR and Stefan use my branch to comment but the PR makes that an easier process    I don't see the dependency error run directly.  I will fix that

taartspi committed:
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

taartspi committed:
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

taartspi committed:
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

taartspi committed:
Work moved into  /pi4j-test/src/main/java/com/pi4j
/test/com/pi4j/test/Main.java (replace contents in Main)
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

taartspi committed:
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

taartspi committed:
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

taartspi committed:
Merge branch 'smokeTestTwo' of https://github.com/Pi4J/pi4j into smokeTestTwo

eitch committed:
[Project] Remove duplicate `pi4j-plugin-ffm` dependency in `pom.xml`.

eitch committed:
[Minor] Clean up unused imports, redundant comments, and improve code formatting in `Main.java`.

eitch committed:
[Tests] Add Raspberry Pi check and improve shutdown handling in `ProviderSmokeTest`
